core/src/main/scala/kafka/log/Log.scala

```diff
@@ -444,6 +457,23 @@ class Log(@volatile var dir: File,
     }
   }
 
+  /*
+   * Increment the log start offset if the provided offset is larger.
+   * And delete old segments whose largest offset < log start offset.
+   */
+  def maybeIncrementLogStartOffsetAndDeleteRecords(offset: Long) {
```

Comment:
@junrao My apology, I missed your comment here. `DeleteRecordsrRequest` can actually be applied on a compacted topic. `DeleteRecordsrRequest` will increase the logStartOffset. When `LogCleaner` calls `log.deleteOldSegments()`, those log segments which breach the logStartOffset will be deleted.